docs/architecture/grpc-for-wcf-developers/interface-definition-language.md

```diff
@@ -1,11 +1,31 @@
 ---
 title: Interface Definition Language - gRPC for WCF Developers
-description: TO BE WRITTEN
+description: Introducing the Protocol Buffers IDL.
 author: markrendle
 ms.date: 09/02/2019
 ---
 
 # Interface Definition Language
 
+With WCF, services can expose description metadata using the Web Service Definition Language (WSDL), which is generated dynamically using .NET Reflection at runtime. Developers can use this metadata to generate clients for those services, potentially in other languages if a platform-neutral binding such as SOAP over HTTP is used.
+
+gRPC uses the Interface Definition Language (IDL) from Protocol Buffers. The Protocol Buffers IDL is a custom, platform-neutral language with an open specification. Developers hand-code `.proto` files to describe services along with their inputs and outputs. These `.proto` files can then be used to generate language- or platform-specific stubs for clients and servers, allowing multiple different platforms to communicate. By sharing `.proto` files, teams can generate code to use each others' services without needing to take a code dependency.
+
+One of the advantages of the Protobuf IDL is that as a custom language it enables gRPC to be completely language and platform agnostic, not favoring any technology over another.
+
+The Protobuf IDL is also much easier for humans to both read and write than WSDL. Changing the WSDL of a WCF service typically requires making the changes to the service code itself, running the service and regenerating the WSDL file from the server. By contrast, with a `.proto` file, changes are simple to apply and automatically flow through the generated code. Visual Studio 2019 builds `.proto` files in the background when they are saved; when using other editors such as VS Code the changes will be applied when the project is built.
+
+When compared with XML, messages encoded using Protobuf have many advantages:
+
+- They can be up to 10x smaller.
```

Comment:
This isn't true when you use WCF's binary message encoding. The difference is only about 10% in size.

Reply:
Just above this (line 18) the doc mentions that these comparisons are to WCF with XML.

I think it is worth discussing either there or below this list that WCF also supports more efficient binary messaging.

Reply:
Added a note.

Reply:
Binary encoding is XML, it's just a binary encoded version of xml.

Reply:
Someone made a blog post and provided code comparing the different serializers and their formats including DataContractSerializer and protobuf among others. I forked their code and added a variant using the binary xml dictionary reader/writer. You can find my fork [here](https://github.com/mconnew/SerializationPerformanceTest_CSharp/tree/BinaryXml) on the BinaryXml branch.  

These are the results:  

**Simple Object**  

| Serializer | Object size | Serialization time | Deserialization time |
| ------------- | ------------- |------------- |------------- |
| DataContractSerializer (text) | 370 | 0.002255ms | 0.005327ms |
| DataContractSerializer (binary) | 83 | 0.000864ms | 0.002067ms |
| Protobuf | 62 | 0.000522ms | 0.001207 ms |

You can see with a simple object, even when using the text xml encoder (which NetTcp doesn't used by default) it's less than 6X bigger. When using the binary xml encoder, protobuf is only 1.33X smaller.

**Array of simple objects**  

| Serializer | Object size | Serialization time | Deserialization time |
| ------------- | ------------- |------------- |------------- |
| DataContractSerializer (text) | 364,299 | 2.431831ms | 4.770156ms |
| DataContractSerializer (binary) | 125,974 | 1.270527ms | 3.068528ms |
| Protobuf | 104,191 | 0.667872ms | 2.445094ms |

When serializing something significantly bigger, there's still nowhere near 10X. For an array of simply objects, when using the text xml encoder it's less than 3.5X bigger. When using the binary xml encoder, protobuf is only 1.21X smaller.

Remember, this is still using the DataContractSerializer so it's still XML, just in a binary form. But even if you are comparing to text, it's a lot less than 10X. I haven't even done anything to try and highlight the strengths of WCF. For example, if I were to enable references with the DataContractSerializer (removes duplicates objects from the output and replaces with a reference) and were to insert the exact same simple object in every position of the array, XML would come out significantly smaller than protobuf as protobuf can't do that.